pkg/cdi/cache.go

```diff
@@ -0,0 +1,275 @@
+// of CDI Spec directories. These can be specified using a WithSpecDirs
+// option. The default set of directories is exposed in DefaultSpecDirs.
+func NewCache(options ...Option) (*Cache, error) {
+	c := &Cache{}
+
+	if err := c.Configure(options...); err != nil {
+		return nil, err
+	}
+	if len(c.specDirs) == 0 {
+		c.Configure(WithSpecDirs(DefaultSpecDirs...))
+	}
+
+	return c, c.Refresh()
+}
+
+// Configure applies options to the cache. Updates the cache if options have
+// changed.
+func (c *Cache) Configure(options ...Option) error {
+	if len(options) == 0 {
+		return nil
+	}
+
+	c.Lock()
+	defer c.Unlock()
+
+	for _, o := range options {
+		if err := o(c); err != nil {
+			return errors.Wrapf(err, "failed to apply cache options")
+		}
+	}
+
+	return nil
+}
+
+// Refresh rescans the CDI Spec directories and refreshes the Cache.
+func (c *Cache) Refresh() error {
+	var (
+		specs      = map[string][]*Spec{}
+		devices    = map[string]*Device{}
+		conflicts  = map[string]struct{}{}
+		specErrors = map[string][]error{}
+		result     []error
+	)
+
+	// collect errors per spec file path and once globally
+	collectError := func(err error, paths ...string) {
+		result = append(result, err)
+		for _, path := range paths {
+			specErrors[path] = append(specErrors[path], err)
+		}
+	}
+	// resolve conflicts based on device Spec priority (order of precedence)
+	resolveConflict := func(name string, dev *Device, old *Device) bool {
+		devSpec, oldSpec := dev.GetSpec(), old.GetSpec()
+		devPrio, oldPrio := devSpec.GetPriority(), oldSpec.GetPriority()
+		switch {
+		case devPrio > oldPrio:
+			return false
+		case devPrio == oldPrio:
+			devPath, oldPath := devSpec.GetPath(), oldSpec.GetPath()
+			collectError(errors.Errorf("conflicting device %q (specs %q, %q)",
+				name, devPath, oldPath), devPath, oldPath)
+			conflicts[name] = struct{}{}
+		}
+		return true
+	}
+
+	_ = scanSpecDirs(c.specDirs, func(path string, priority int, spec *Spec, err error) error {
+		path = filepath.Clean(path)
+		if err != nil {
+			collectError(errors.Wrapf(err, "failed to load CDI Spec"), path)
+			return nil
+		}
+
+		vendor := spec.GetVendor()
+		specs[vendor] = append(specs[vendor], spec)
+
+		for _, dev := range spec.devices {
+			qualified := dev.GetQualifiedName()
+			other, ok := devices[qualified]
+			if ok {
+				if resolveConflict(qualified, dev, other) {
+					continue
+				}
+			}
+			devices[qualified] = dev
+		}
+
+		return nil
+	})
+
+	for conflict := range conflicts {
+		delete(devices, conflict)
+	}
+
+	c.Lock()
+	defer c.Unlock()
+
+	c.specs = specs
+	c.devices = devices
+	c.errors = specErrors
+
+	if len(result) > 0 {
+		return multierror.Append(nil, result...)
+	}
+
+	return nil
+}
+
+// InjectDevices injects the given qualified devices to an OCI Spec. It
+// returns any unresolvable devices and an error if injection fails for
+// any of the devices.
+func (c *Cache) InjectDevices(ociSpec *oci.Spec, devices ...string) ([]string, error) {
```

Comment:
Does it really make sense to have `InjectDevices` associated with a "cache"? The storage of the specs and applying the required OCI modifications seem like different (orthogonal) concerns.

Reply:
Well, the whole idea here is that if you are a consumer of CDI specs (a runtime or the orchestration), then you
  - only need to deal with the cache/registry,
  - can perform commonly needed tasks (injection) with preferably a single function call, and
  - without having to understand any intricacies of CDI itself

About deduplication of devices to inject vs. deduplication of global-scope container edits from CDI Specs... The latter is an intricacy of CDI itself, so we need/want to take care of it. AFAICT, this is different from deduplicating the consumer-provided set of devices to inject. While that would definitely be also possible, unless there is a common use case where CDI duplicates in the input naturally arise/occur without special care being taken by the consumer, I think we should just process the input as is without trying to filter duplicates (and let requests with duplicates eventually fail in the worst case, if that happens to be the runtime's behavior).